spec/lib/services/nodes-api-service-spec.js

```diff
@@ -73,88 +40,84 @@ describe("Http.Services.Api.Nodes", function () {
         });
 
-        it("removeNode should fail when a workflow is running", function() {
-            this.sandbox.stub(taskGraphProtocol, 'getActiveTaskGraph').resolves('true');
-            return expect(nodeApiService.removeNode(computeNode))
-                   .to.be.rejectedWith(Errors.BadRequestError);
-        });
-
-        it("removeNode should remove one node", function() {
-            this.sandbox.stub(taskGraphProtocol, 'getActiveTaskGraph').resolves('');
-            this.sandbox.stub(nodeApiService, '_findEnclNodes').resolves();
-            this.sandbox.stub(nodeApiService, '_removeEnclRelations').resolves();
+        it("_findTargetNodes should find related enclosure nodes", function() {
+            needByIdentifier.resolves(enclosureNode);
 
-            return nodeApiService.removeNode(computeNode)
-            .then(function (node) {
-                expect(waterline.lookups.update).to.have.been.calledOnce;
-                expect(waterline.nodes.destroy).to.have.been.calledOnce;
-                expect(waterline.catalogs.destroy).to.have.been.calledOnce;
-                expect(waterline.workitems.destroy).to.have.been.calledOnce;
-                expect(nodeApiService._findEnclNodes).to.have.been.calledOnce;
-                expect(nodeApiService._removeEnclRelations).to.have.been.calledOnce;
-                expect(node).to.equal(computeNode);
-            });
-        });
-    });
-    describe("_findEnclNodes", function() {
-        before("_findEnclNodes before", function() {
-        });
-
-        beforeEach(function() {
-        });
-
-        it("_findEnclNodes should find related enclosure nodes", function() {
-            this.sandbox.stub(waterline.nodes, 'needByIdentifier').resolves(enclosureNode);
-
-            return nodeApiService._findEnclNodes(computeNode)
+            return nodeApiService._findTargetNodes(computeNode, 'enclosedBy')
             .then(function (nodes) {
-                expect(waterline.nodes.needByIdentifier).to.have.been.calledOnce;
+                expect(needByIdentifier).to.have.been.calledOnce;
                 expect(nodes[0]).to.equal(enclosureNode);
             });
         });
 
-        it("_findEnclNodes should return nothing if cannot find enclosure node", function() {
-            this.sandbox.stub(waterline.nodes, 'needByIdentifier').rejects(Errors.NotFoundError(''));
+        it("_findTargetNodes should return nothing if cannot find enclosure node", function() {
+            needByIdentifier.rejects(Errors.NotFoundError(''));
 
-            return nodeApiService._findEnclNodes(computeNode)
+            return nodeApiService._findTargetNodes(computeNode, 'enclosedBy')
             .then(function (nodes) {
-                expect(waterline.nodes.needByIdentifier).to.have.been.calledOnce;
+                expect(needByIdentifier).to.have.been.calledOnce;
                 expect(nodes[0]).to.equal(undefined);
```

Comment:
If target node cannot be found, I think return empty array is better if you don't want throw error. Because I see you have a lot of Promise.map for the returned target node (undefined) in your code.